handwriting_line_recognition.py

```diff
@@ -0,0 +1,392 @@
+import time
+import random
+import os
+import matplotlib.pyplot as plt
+import argparse
+import string
+
+import mxnet as mx
+import numpy as np
+from skimage import transform as skimage_tf
+from skimage import exposure
+
+from mxnet import nd, autograd, gluon
+from mxboard import SummaryWriter
+from mxnet.gluon.model_zoo.vision import resnet34_v1
+np.seterr(all='raise')
+
+import multiprocessing
+mx.random.seed(1)
+
+from utils.iam_dataset import IAMDataset
+
+max_seq_len = 100
+print_every_n = 5
+save_every_n = 50
+send_image_every_n = 10
+
+from utils.iam_dataset import IAMDataset
+from utils.draw_text_on_image import draw_text_on_image
+
+alphabet_encoding = string.ascii_letters+string.digits+string.punctuation+' '
+alphabet_dict = {alphabet_encoding[i]:i for i in range(len(alphabet_encoding))}
+
+class EncoderLayer(gluon.Block):
+    '''
+    The encoder layer takes the image features from a CNN. The image features are transposed so that the LSTM 
+    slices of the image features can be sequentially fed into the LSTM from left to right (and back via the
+    bidirectional LSTM). 
+    '''
+    def __init__(self, hidden_states=200, lstm_layers=1, **kwargs):
+        super(EncoderLayer, self).__init__(**kwargs)
+        with self.name_scope():
+            self.lstm = mx.gluon.rnn.LSTM(hidden_states, lstm_layers, bidirectional=True)
+            
+    def forward(self, x):
+        x = x.transpose((0, 3, 1, 2))
+        x = x.flatten()
+        x = x.split(num_outputs=max_seq_len, axis=1) # (SEQ_LEN, N, CHANNELS)
+        x = nd.concat(*[elem.expand_dims(axis=0) for elem in x], dim=0)
+        x = self.lstm(x)
+        x = x.transpose((1, 0, 2)) #(N, SEQ_LEN, HIDDEN_UNITS)
+        return x
+
+class Network(gluon.Block):
+    def __init__(self, num_downsamples=2, resnet_layer_id=4, lstm_hidden_states=200, lstm_layers=1, **kwargs):
+        super(Network, self).__init__(**kwargs)
+        self.p_dropout = 0.5
+        self.num_downsamples = num_downsamples
+        self.body = self.get_body(resnet_layer_id=resnet_layer_id)
+
+        self.encoders = gluon.nn.Sequential()
+        
+        for _ in range(self.num_downsamples):
+            encoder = self.get_encoder(lstm_hidden_states=lstm_hidden_states, lstm_layers=lstm_layers)
+            self.encoders.add(encoder)
+        self.decoder = self.get_decoder()
+        self.downsampler = self.get_down_sampler(64)
```

Comment:
make 64 a parameter

Reply:
This is kind of a constant since the resnet feature extractor has 64 filters.

Reply:
I see, can you name it`FEATURE_EXTRACTOR_FILTER=64` ?